Bluetooth: Mesh: Let vendor models receive messages with SIG opcodes

ludvigsj opened the pull request:
This fixes a limitation where the stack assumed that vendor models could
not specify handlers for SIG-defined (1- or 2-byte) opcodes. This
assumption does not exist in the specification. In fact, the
specification expects this very use-case to be possible, in MeshPRT 1.1,
section 3.8.3:

> To exchange messages with a Bluetooth SIG adopted model, a Vendor Model
> shall use the Access message defined for the Bluetooth SIG model.

In addition, it includes a commit to add a BSIM test to check this
behvior, a commit to fix a compile warning when not forcing CID to match
between model and opcode handlers, and a commit to deprecate the forced CID
match (and related optimization) option and disable it by default. See individual commit messages
for details.

ludvigsj committed:
Bluetooth: Mesh: Fix compile warning when CID check is disabled

Since the entire cid checking function is conditionally included based
on KConfig, the usage site also needs to use #if instead of IS_ENABLED,
otherwise you get the following warning during compilation when the
option is disabled:

`error: implicit declaration of function
‘bt_mesh_vnd_mod_msg_cid_check’`

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>

ludvigsj committed:
Bluetooth: Mesh: Let vendor models receive messages with SIG opcodes

This fixes a limitation where the stack assumed that vendor models could
not specify handlers for SIG-defined (1- or 2-byte) opcodes. This
assumption does not exist in the specification. In fact, the
specification expects this very use-case to be possible, in MeshPRT 1.1,
section 3.8.3:

> To exchange messages with a Bluetooth SIG adopted model, a Vendor Model
> shall use the Access message defined for the Bluetooth SIG model.

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>

ludvigsj committed:
Bluetooth: Mesh: Test that vendor models can receive SIG opcodes

This adds a test case to the access BSIM tests to check that vendor
models can specify a handler for and receive messages using a
SIG-defined 1-byte opcode.

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>

omkar3141 commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1388,36 +1387,16 @@ static bool model_has_dst(const struct bt_mesh_model *mod, uint16_t dst, const u
-static const struct bt_mesh_model_op *find_op(const struct bt_mesh_elem *elem,
-					      uint32_t opcode, const struct bt_mesh_model **model)
+static const struct bt_mesh_model_op *find_op_in_list(const struct bt_mesh_model *models,
+						      uint8_t count, uint32_t opcode,
```
minor
```suggestion
						      uint8_t mdl_count, uint32_t opcode,
```
?

ludvigsj commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1434,6 +1413,31 @@ static const struct bt_mesh_model_op *find_op(const struct bt_mesh_elem *elem,
+	/* Return (NULL) already if we force vendor model opcode CID match. SIG opcodes per
+	 * definition cannot match since they don't have a CID.
+	 */
+	if (op != NULL || IS_ENABLED(CONFIG_BT_MESH_MODEL_VND_MSG_CID_FORCE)) {
```
`find_op_in_list` will loop through every vendor model before making this decision. Since `CONFIG_BT_MESH_MODEL_VND_MSG_CID_FORCE` is supposed to optimize I thought it would be better to just return early here.

ludvigsj replied:
Oh, and "if opcode is 3 byte long", no, this line only runs when opcode is 1-2 bytes long. The `if` in the beginning of the function is the only code that runs if it's 3 bytes long.

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1388,36 +1387,16 @@ static bool model_has_dst(const struct bt_mesh_model *mod, uint16_t dst, const u
-		     cid != UINT32_MAX &&
-		     cid != models[i].vnd.company) {
+		    BT_MESH_MODEL_OP_LEN(opcode) == 3 &&
+		    (uint16_t)(opcode & 0xffff) != models[i].vnd.company) {
```
I don't like this tbh. Passing here sig models as `models` but accessing `company` field of vendor models. Even though it is currently safe because [company](https://github.com/zephyrproject-rtos/zephyr/blob/c3190d34aa912d39cd082db822f6ebb332806a2b/include/zephyr/bluetooth/mesh/access.h#L707) and [id](https://github.com/zephyrproject-rtos/zephyr/blob/c3190d34aa912d39cd082db822f6ebb332806a2b/include/zephyr/bluetooth/mesh/access.h#L716) are the same types at the same location, it may change and then this code causes undefined behavior. Neither logically I understand this code.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_access.c:
```diff
@@ -98,6 +103,7 @@ static const struct {
 };
 
 static struct k_sem publish_sem;
+static struct k_sem vnd_rx_sem;
```
I think both semaphores should be statically initialized, otherwise it will be undefined behavior if some test use wrong message triggering a sempahore which is not initialized.

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -999,7 +999,7 @@ static int bt_mesh_vnd_mod_msg_cid_check(const struct bt_mesh_model *mod)
 		cid = (uint16_t)(op->opcode & 0xffff);
 
-		if (cid == mod->vnd.company) {
+		if (BT_MESH_MODEL_OP_LEN(op->opcode) == 3 && cid == mod->vnd.company) {
```
I don't think this is correct since the force option is to exactly ensure that a vendor model doesn't serve messages opcode with different Company ID.

Since the spec clearly allows a vendor model serve 1-/2-octets opcodes, this should only check matching Company ID for 3-octets opcode messages. Therefore, I think the check should be like this:
```
if (BT_MESH_MODEL_OP_LEN(op->opcode) == 3 && cid != mod->vnd.company) {
    LOG_ERR("...");
    return -EINVAL;
}
```

ludvigsj commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -785,11 +785,14 @@ config BT_MESH_ACCESS_LAYER_MSG
-	bool "Force vendor model to use the corresponding CID field message"
-	default y
+	bool "Force vendor model to use opcodes with matching CID"
+	default n
```
Because

1. being able to use opcodes that don't match the CID of the model is a very normal behavior, and having this fail is already confusing users who don't know about this option. It is even required by the spec to be able to communicate with other models, see commit message.
2. The change in behavior is slight, I would be surprised if anyone was consciously relying on this optimization. Note, every application that compiles and runs properly will still do so after this change, the _only_ consequence will be a slight performance decrease for applications with multiple vendor models using at least two different CIDs, which is such a niche use case that I decided that it didn't warrant a migration guide.
3. Since the corner case where this optimization can do anything at all (at least two different vendor models with at least two different CIDs in the network), and the benefit of it when it _does_ apply (saving a tiny handful of comparisons per opcode defined in vendor models on a node) I'd prefer to just remove the entire thing.

ludvigsj replied:
Regarding 2), I'm not even sure if this is true; my extra `if`s and function calls might negate that already, meaning even without changing the deafult they get decreased performance.

ludvigsj replied:
We discussed this in the Mesh team, and we really don't like this defaulting to y and confusing users who want to just follow what the spec says is allowed (and sometimes required!) and use other opcodes in their models.

PavelVPV replied:
Your comments are correct with the assumption of the new change from this PR. This is not correct if the change is such that this Kconfig option only verifies 3-octet opcodes with CID (like before) leaving 1-/2-octet opcodes ignored, thus making the code spec compliant even if it is enabled. Right? I guess the initial intention of the option was exactly to have opcode.CID with model.CID match, disabling which you can build a hacked model.

ludvigsj commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -785,11 +785,14 @@ config BT_MESH_ACCESS_LAYER_MSG
-	  corresponding CID field.
+	  This option forces vendor models to only use opcodes with a CID field
+	  matching the models CID, making it possible for the stack to slightly
+	  optimize message handling in applications that use vendor models, where
```
I agree, I'm not sure if it does optimize anything. We are probably going to deprecate and remove this entire thing if we find the time, it looks a bit useless and complicates the code.

ludvigsj replied:
The spec does not prohibit you from using opcodes with CID1 in a model with CID2. This is an artificial limitation we introduce.

jhedberg commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1388,36 +1387,16 @@ static bool model_has_dst(const struct bt_mesh_model *mod, uint16_t dst, const u
-		     cid != UINT32_MAX &&
-		     cid != models[i].vnd.company) {
+		    BT_MESH_MODEL_OP_LEN(opcode) == 3 &&
+		    (uint16_t)(opcode & 0xffff) != models[i].vnd.company) {
```
@ludvigsj that sounds fine to me

ludvigsj committed:
Bluetooth: Mesh: Deprecate VND_MSG_CID_FORCE option

This deprecates CONFIG_BT_MESH_MODEL_VND_MSG_CID_FORCE. This option was
added to allow an optimization that a) only can apply in edge case
networks that use multiple vendor models with different CIDs, b) doesn't
really have much potential gains even in this case unless there is a
large number of opcodes in these vendor models, and c) makes it
impractical to implement support for vendor models handling SIG opcodes.

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>

PavelVPV commented on tests/bsim/bluetooth/mesh/tests_scripts/access/access_vnd_rcv_sig_msg.sh:
```diff
@@ -0,0 +1,14 @@
+
+source $(dirname "${BASH_SOURCE[0]}")/../../_mesh_test.sh
+
+# Test scenario:
```
Please add purpose of the test.

PavelVPV commented:
> Hi @ludvigsj, @PavelVPV, could you elaborate why was deprecated option mentioned in migration guide but not in the release notes?
> 
> There is already mentioned deprecated API related to OOB authentication: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-4.4.rst?plain=1#L72-L79
> 
> Seems kind of inconsistent to me. To figure out full set of deprecated functionality customer needs to read 2 documents.

https://docs.zephyrproject.org/latest/releases/index.html#migration-guides

I think it makes sense to mention the deprecation in the migration guide since the option has no effect anymore and the second sentence clarifies that. If you disagree, request a change with a reason and suggestion.

I agree that it should be mentioned in the release notes as well due to this:
https://github.com/zephyrproject-rtos/zephyr/blob/7963b69fa451c6b611040597e6d99e0dc3c61be2/doc/releases/release-notes-4.4.rst?plain=1#L4

omkar3141 commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -437,6 +280,12 @@ Bluetooth Host
+==============
+
+* :kconfig:option:`CONFIG_BT_MESH_MODEL_VND_MSG_CID_FORCE` has been deprecated. Enabling it no
+  longer has any effect on message handling performance.
```
Ok. But 'enabling it will cause vendor model initialization to fail, should two different vendor models share the opcodes'. This fact will become known only at run time when user does this and it is not mentioned anywhere. It is okay to keep the code, but can you mention this fact somewhere? I think it is better to mention here, because anyone dependent on this option will read this part of migration guide and become aware of consequences?

alxelax reviewed (changes requested):
@ludvigsj, please update release notes with deprecation option too (see reference in my initial message).
